man/man1/tabdist.1
.TH TABDIST 1NEMO "11 June 2026"

.SH "NAME"
tabdist \- Compare distances between points in two tables, aka catalog matching

.SH "SYNOPSIS"
\fBtabdist\fP [parameter=value]

.SH "DESCRIPTION"
Given two tables, each containing specified column(s) that represent a set
of 1D, 2D or 3D points, it will
compute the nearest neigbor(s) from each point in the first table
to those from the second table. 
.PP
The output contains the same number of rows as the input number of
rows in the first dataset.
.PP
There is also a hack to find the angular distance on the sky between two points.
See examples below.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin1=\fP
First dataset, an ascii table.
.br
No default.
.TP
\fBin2=\fP
Second dataset, an ascii table. 
The tables do not need to be of the same length.
.br
No default.
.TP
\fBcol1=\fP
Coordinate columns from 1st dataset. 1D, 2D and 3D are currently supported
.br
[1,2]
.TP
\fBcol2=\fP
Coordinate columns from 2nd dataset. 1D, 2D and 3D are currently supported
.br
[1,2]
.TP
\fBid1=\fP
Column in first dataset representing the ID name.
By default a number between 1 and number of sources is used.
.br
[0]
.TP
\fBid2=\fP
Column in second dataset representing the ID name
.br
[0]
.TP
\fBradec=t|f\fP
Should columns be interpreted as angular (usually RA/DEC)?
Only 2D comparisions are allowed here. Units must be radians at the moment.
See also \fItabdms(1NEMO)\fP.
.br
Default: false
.TP
\fBunits=r|d\fP
Units of the X,Y angular coordinates, radians or degrees.
.br
Default: degree
.TP
\fBdmin=\fP
Only print distances less than this value.   Units: arcsec if input units=degree

.SH "EXAMPLES"
The X,Y,Z coordinates of a 3 particle plummer sphere are perturbed by a small
amount and then compared to the original list. Thus they must match one on one:
.nf
% mkplummer - 3 | snapprint - x,y,z > tab11
% tabmath tab11 - '%1+rang(0,0.01),%2+rang(0,0.01),%3+rang(0,0.01)' > tab12
.fi

First a one dimensional comparison

.nf
% tabdist tab12 tab12 1 4
1  0.117073   1  0.122846     0.005773
2 -0.00899587 2 -0.0100997    0.00110383
3 -0.108077   3 -0.0964223    0.0116547
.fi
and then a full 3D case:
.nf
% tabdist tab12 tab12 1,2,3 4,5,6
1  0.117073  -0.097704 -0.299081   1  0.122846  -0.102288 -0.317648   0.0199768
2 -0.00899587 0.293374  0.12652    2 -0.0100997  0.281876  0.121453   0.0126134 
3 -0.108077  -0.19567   0.172561   3 -0.0964223 -0.203288  0.17993    0.0157534

.fi
Currently it will display the line number, the coordinate(s) in table1,
then the best matching line number and coordinates(s) from table2, and finally the distance.

.PP
To compute the angular distance between two points is a bit kludgy.  Put the two sources
in two tables as follows:
.EX
   $ tabdist tab1 tab2 2,3 2,3 radec=t dmin=9999
   VIIZW700 258.847083 57.431389 258.847 57.4314    NGC6338  258.845812 57.411185 258.846 57.4112    72.7761

   $ cat tab1 tab2
   VIIZW700 258.847083 57.431389 9539
   NGC6338  258.845812 57.411185 8185
.EE
the two galaxies are 73 arcsec apart.

.SH "CAVEATS"
Currently a brute force (quadratic in N) minimum distance is used.

.SH "SEE ALSO"
tabmath(1NEMO), tabdms(1NEMO), comparecats(SIRTF.map2)
.PP
k-d Match: http://arxiv.org/abs/1304.0838

.SH "FILES"
src/kernel/tab/tabdist.c

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
13-Feb-13	V0.1 Created	PJT
13-nov-2013	V0.3 added radec=	PJT
18-nov-2013	V0.4 added id1,id2=	PJT
1-apr-2026	V0.8 ansi fixes, print full line	PJT
.fi


src/kernel/tab/tabdist.c
/*
 *  Compare distance between points in two tables
 *
 *      13-feb-2013     0.1    Created, Q&D
 *      24-jun-2016     0.7    allow radec in degrees; fix C language issues, introduce dmin=,
 *                             faster precompute
 *       1-apr-2026     0.8    ansi compiler + cleanup
 *
 *  See also the SIRTF/map2 project in 'c2d'
 *
 *  @todo   consider allowing scaling/offset between two datasets 
 *          Y = aX + b
 *          use k-d Match?   http://arxiv.org/abs/1304.0838
 *  @todo   ID column
 *
 *  @todo   Faster lookup using 2D grid with linked list?
 *
 *  @todo   Generalize to N tables?
 *
 *  @todo   Pick if to use short lines (id,coord) or full lines from input table
 */

#include <stdinc.h>
#include <getparam.h>
#include <extstring.h>
#include <grid.h>
#include <table.h>

string defv[] = {
    "in1=???\n		First (ascii table) dataset",
    "in2=???\n		Second (ascii table) dataset",
    "col1=1,2\n         X (optionally Y and Z) Column from 1st dataset",
    "col2=1,2\n         X (optionally Y and Z) Column from 2nd dataset",
    "id1=0\n            Column in first data reprenting the ID",
    "id2=0\n            Column in second data reprenting the ID",
    "radec=false\n      Interpret X and Y as RA/DEC angles on sky",
    "units=d\n          radians or degree for 2D case in case of radec=t",
    "xgrid=\n           Grid to limit nearest neighbors for special 2D case",
    "ygrid=\n           Grid to limit nearest neighbors for special 2D case",
    "dmin=-1\n          If positive, only print distances less than this value",
    "all=t\n            Show all columns of match, or just the coords",
    "VERSION=0.9\n	2-apr-2026 PJT",
    NULL,
};

string usage="Compare distances between points in two tables";

//#define MAXGRID  1024

void compare_1d (int npt1, int npt2, real *x1, real *x2, real dmin);
void compare_2d (int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real dmin);
void compare_2da(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real scale, string *name1, string *name2, real dmin);
void compare_2da_slow(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real s, string *name1, string *name2, real dmin_out);
void compare_2dg(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, Grid *g, real dmin);
void compare_3d (int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real *z1, real *z2, real dmin);

void get_atable_a(stream instr, int id, int npt, string *name);
void get_atable_l(stream instr, int npt, string *name);

  
inline static real dist1(real x1,real x2) {
  real d = x1-x2;
  if (d > 0.0) return d;
  return -d;
}

inline static real dist2(real x1,real y1,real x2,real y2) {
  return (x1-x2)*(x1-x2)+(y1-y2)*(y1-y2);
}

inline static real dist3(real x1,real y1,real z1,real x2,real y2,real z2) {
  return (x1-x2)*(x1-x2)+(y1-y2)*(y1-y2)+(z1-z2)*(z1-z2);
}

/* slow version : compute sin() and cos() in the fly */

inline static real dist2a(real x1,real y1,real x2,real y2) {
  return sin(y1)*sin(y2)+cos(y1)*cos(y2)*cos(x1-x2);
}

/* fast version : stores and precomputes */
   
static real *X1, *X2;
static real *Sinx1, *Sinx2, *Siny1, *Siny2;
static real *Cosx1, *Cosx2, *Cosy1, *Cosy2;

inline static real dist2ai(int i,int j) {
  return Siny1[i]*Siny2[j] + Cosy1[i]*Cosy2[j]*cos(X1[i]-X2[j]);
}



void key2grid(string key, Grid *g)
{

}


#define NDIM   3

void nemo_main()
{
  int i, npt1, npt2, ncol1, ncol2, id1, id2;
  real *coldat1[NDIM], *coldat2[NDIM];
  int colnr1[NDIM], colnr2[NDIM];
  real *x1 = NULL, *x2 = NULL;
  real *y1 = NULL, *y2 = NULL;
  real *z1 = NULL, *z2 = NULL;
  real ascale = 1.0;
  real dmin = getrparam("dmin");
  string input1 = getparam("in1");
  string input2 = getparam("in2");
  stream instr1, instr2;
  string *name1, *name2, *line1, *line2, *show1, *show2;
  bool Qradec = getbparam("radec");
  bool Qall = getbparam("all");
  string units = getparam("units");
  Grid gx, gy;

  if (Qradec) warning("new radec mode ");
  if (*units == 'd') ascale = PI/180.0;

  ncol1 = nemoinpi(getparam("col1"),colnr1,NDIM);
  ncol2 = nemoinpi(getparam("col2"),colnr2,NDIM);

  key2grid("xgrid",&gx);
  key2grid("ygrid",&gy);

  if (ncol1 != ncol2) error("column count error: ncol1=%d and ncol2=%d not same",ncol1,ncol2);
  if (ncol1 < 1 || ncol1 > NDIM) error("ncol1=%d not supported",ncol1);
  
  instr1 = stropen(input1,"r");
  instr2 = stropen(input2,"r");

  npt1 = nemo_file_lines(input1,0);
  npt2 = nemo_file_lines(input2,0);

  if (ncol1 > 0) {
    x1 = (real *) allocate(npt1 * sizeof(real));
    x2 = (real *) allocate(npt2 * sizeof(real));
  }
  if (ncol1 > 1) {
    y1 = (real *) allocate(npt1 * sizeof(real));
    y2 = (real *) allocate(npt2 * sizeof(real));
  }
  if (ncol1 > 2) {
    z1 = (real *) allocate(npt1 * sizeof(real));
    z2 = (real *) allocate(npt2 * sizeof(real));
  }
  
  coldat1[0] = x1;     coldat2[0] = x2;   
  coldat1[1] = y1;     coldat2[1] = y2;   
  coldat1[2] = z1;     coldat2[2] = z2;   

  npt1 = get_atable(instr1,ncol1,colnr1,coldat1,npt1);
  npt2 = get_atable(instr2,ncol2,colnr2,coldat2,npt2);

  dprintf(0,"Npt1=%d Npt2=%d\n",npt1,npt2);

  if (Qradec) {   /* fast precompute for angular case */
    warning("precompute, ascale=%g",ascale);
    X1 = (real *) allocate(npt1 * sizeof(real));
    X2 = (real *) allocate(npt2 * sizeof(real));
    Sinx1 = (real *) allocate(npt1 * sizeof(real));
    Siny1 = (real *) allocate(npt1 * sizeof(real));
    Cosx1 = (real *) allocate(npt1 * sizeof(real));
    Cosy1 = (real *) allocate(npt1 * sizeof(real));
    Sinx2 = (real *) allocate(npt2 * sizeof(real));
    Siny2 = (real *) allocate(npt2 * sizeof(real));
    Cosx2 = (real *) allocate(npt2 * sizeof(real));
    Cosy2 = (real *) allocate(npt2 * sizeof(real));
    for (i=0; i<npt1; i++) {
      X1[i]    = x1[i]*ascale;
      Sinx1[i] = sin(x1[i]*ascale);
      Siny1[i] = sin(y1[i]*ascale);
      Cosx1[i] = cos(x1[i]*ascale);
      Cosy1[i] = cos(y1[i]*ascale);
    }
    for (i=0; i<npt2; i++) {
      X2[i]    = x2[i]*ascale;
      Sinx2[i] = sin(x2[i]*ascale);
      Siny2[i] = sin(y2[i]*ascale);
      Cosx2[i] = cos(x2[i]*ascale);
      Cosy2[i] = cos(y2[i]*ascale);
    }
  }

  id1 = getiparam("id1");
  id2 = getiparam("id2");
  if (id1 > 0) {
    rewind(instr1);
    name1 = (string *) allocate(npt1 * sizeof(string));
    get_atable_a(instr1,id1,npt1,name1);
  } else
    name1 = NULL;
  if (id2 > 0) {
    rewind(instr2);
    name2 = (string *) allocate(npt2 * sizeof(string));
    get_atable_a(instr2,id2,npt2,name2);
  } else
    name2 = NULL;

  if (Qall) {
    rewind(instr1);
    line1 = (string *) allocate(npt1 * sizeof(string));
    get_atable_l(instr1,npt1,line1);

    rewind(instr2);
    line2 = (string *) allocate(npt2 * sizeof(string));
    get_atable_l(instr2,npt2,line2);
    show1 = line1;
    show2 = line2;
  } else {
    show1 = name1;
    show2 = name2;
  }
  
  strclose(instr1);
  strclose(instr2);

  if (Qradec && ncol1==2)
    compare_2da_slow(npt1, npt2, x1, x2, y1, y2, ascale, show1, show2, dmin);
  else if (ncol1 == 1) 
    compare_1d(npt1, npt2, x1, x2, dmin);
  else if (ncol1 == 2) 
    compare_2d(npt1, npt2, x1, x2, y1, y2, dmin);
  else if (ncol1 == 3)
    compare_3d(npt1, npt2, x1, x2, y1, y2, z1, z2, dmin);
  else
    warning("%d columns not supported",ncol1);
}


void compare_1d(int npt1, int npt2, real *x1, real *x2, real dmin_out)
{
  int i, j, jmin;
  real d, dmin;
  int count=0;
  
  for (i=0; i<npt1; i++) {
    dmin = dist1(x1[i],x2[0]);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist1(x1[i],x2[j]);
      if (d<dmin) {
	dmin = d;
	jmin = j;
      }
    }
    if (dmin_out>0 && dmin<dmin_out) {
      count++;
      printf("%d %g   %d %g   %g\n",i+1,x1[i],jmin+1,x2[jmin], dmin);
    }
  }
  printf("# Found %d matches for dmin=%g in 1d\n",count,dmin_out);    
}

void compare_2d(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real dmin_out)
{  
  int i, j, jmin;
  real d, dmin;
  int count=0;
  
  for (i=0; i<npt1; i++) {
    dmin = dist2(x1[i],y1[i],x2[0],y2[0]);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist2(x1[i],y1[i],x2[j],y2[j]);
      if (d<dmin) {
	dmin = d;
	jmin = j;
      }
    }
    dmin = sqrt(dmin);
    if (dmin_out>0 && dmin<dmin_out) {
      count++;
      printf("%d %g %g   %d %g %g   %g\n",i+1,x1[i],y1[i],jmin+1,x2[jmin], y2[jmin],dmin);
    }
  }
  printf("# Found %d matches for dmin=%g in 2d\n",count,dmin_out);  
}

/* fast version - doesn't work yet 
 * Npt1=1493 Npt2=12681
 * cartesian:  0.2"
 * slow:       8.0"
 * fast:       2.1"
 */

/* fast */
void compare_2da(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real s, string *name1, string *name2, real dmin_out)
{  
  int i, j, jmin;
  real d, dmax, dmin;
  string n1, n2;
  char ord1[16], ord2[16];

  
  for (i=0; i<npt1; i++) {
    if (name1) dprintf(2,"NAME1: %s\n",name1[i]);
    dmax = dist2ai(i,0);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist2ai(i,j);
      if (d>dmax) {
	dmax = d;
	jmin = j;
      }
    }
    dmin = acos(dmax)*DR2AS;  /* radians to arcsec */

    if (name1==NULL) {
      sprintf(ord1,"%d",i+1);
      n1 = ord1;
    } else
      n1 = name1[i];

    if (name2==NULL) {
      sprintf(ord2,"%d",jmin+1);
      n2 = ord2;
    } else
      n2 = name2[jmin];

    if (dmin_out>0 && dmin<dmin_out)
      printf("%s %g %g   %s %g %g   %g\n",n1,x1[i],y1[i],n2,x2[jmin],y2[jmin],dmin);
  }
}

/* slow */
void compare_2da_slow(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real s, string *name1, string *name2, real dmin_out)
{  
  int i, j, jmin;
  real d, dmax, dmin;
  string n1, n2;
  char ord1[16], ord2[16];
  int count = 0;

  
  for (i=0; i<npt1; i++) {
    if (name1) dprintf(2,"NAME1: %s\n",name1[i]);
    dmax = dist2a(x1[i]*s,y1[i]*s,x2[0]*s,y2[0]*s);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist2a(x1[i]*s,y1[i]*s,x2[j]*s,y2[j]*s);
      if (d>dmax) {
	dmax = d;
	jmin = j;
      }
    }
    dmin = acos(dmax)*DR2AS;  /* radians to arcsec */

    if (name1==NULL) {
      sprintf(ord1,"%d",i+1);
      n1 = ord1;
    } else
      n1 = name1[i];

    if (name2==NULL) {
      sprintf(ord2,"%d",jmin+1);
      n2 = ord2;
    } else
      n2 = name2[jmin];

    if (dmin_out>0 && dmin<dmin_out) {
      count++;
      //printf("%s %s   %g\n",name1[i],name2[jmin],dmin);      
      printf("%s %g %g    %s %g %g    %g\n",n1,x1[i],y1[i],n2,x2[jmin],y2[jmin],dmin);
    }
  }
  printf("# Found %d matches for dmin=%g in 2da\n",count,dmin_out);
}

void compare_2g(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, Grid *gx, Grid *gy, real dmin_out)
{  
  int i, j, jmin;
  real d, dmin;
  int count = 0;
  
  for (i=0; i<npt1; i++) {
    dmin = dist2(x1[i],y1[i],x2[0],y2[0]);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist2(x1[i],y1[i],x2[j],y2[j]);
      if (d<dmin) {
	dmin = d;
	jmin = j;
      }
    }
    dmin = sqrt(dmin);
    if (dmin_out>0 && dmin<dmin_out) {
      count++;
      printf("%d %g %g   %d %g %g   %g\n",i+1,x1[i],y1[i],jmin+1,x2[jmin], y2[jmin],dmin);
    }
  }
  printf("# Found %d matches for dmin=%g in 2g\n",count,dmin_out);
}


void compare_3d(int npt1, int npt2, real *x1, real *x2, real *y1, real *y2, real *z1, real *z2, real dmin_out)
{  
  int i, j, jmin;
  real d, dmin;
  int count = 0;
  
  for (i=0; i<npt1; i++) {
    dmin = dist3(x1[i],y1[i],z1[i],x2[0],y2[0],z2[0]);
    jmin = 0;
    for (j=1; j<npt2; j++) {
      d = dist3(x1[i],y1[i],z1[i],x2[j],y2[j],z2[j]);
      if (d<dmin) {
	dmin = d;
	jmin = j;
      }
    }
    dmin = sqrt(dmin);
    if (dmin_out>0 && dmin<dmin_out) {
      count++;
      printf("%d %g %g %g   %d %g %g %g   %g\n",i+1,x1[i],y1[i],z1[i],jmin+1,x2[jmin],y2[jmin],z2[jmin],dmin);
    }
  }
  printf("# Found %d matches for dmin=%g in 3d\n",count,dmin_out);
}

/* 
 *  read lines and pick a column as the ID.
 *  skip lines that start with '#'
 */

void get_atable_a(stream instr, int id, int npt, string *name)
{
  static char line[MAX_LINELEN];
  string *sp;
  int i;

  for (i=0; i<npt; ) {
    if (fgets(line,MAX_LINELEN,instr) == NULL)
      error("unexpected EOF on table read (%d,%d)",id,npt);
    if (line[0] == '#') continue;
    sp = burststring(line,", \t\r");
    if (xstrlen(sp,sizeof(string))<id-1) {
      warning("skipping line, not enough columns");
      continue;
    }
    name[i] = strdup(sp[id-1]);
    freestrings(sp);
    i++;
  }
  dprintf(1,"First and last ID: %s %s\n",name[0],name[npt-1]);
}

void get_atable_l(stream instr, int npt, string *name)
{
  static char line[MAX_LINELEN];
  int i;

  for (i=0; i<npt; ) {
    if (fgets(line,MAX_LINELEN,instr) == NULL)
      error("unexpected EOF on table read (%d)",npt);
    if (line[0] == '#') continue;
    line[strlen(line)-1] = '\0';
    name[i] = strdup(line);
    i++;
  }
  dprintf(1,"First and last ID: %s %s\n",name[0],name[npt-1]);
}
